R/derive_extreme_event.R

```diff
@@ -102,56 +148,290 @@
+#'       set_values_to = exprs(AVALC = "Y")
+#'     )
+#'   ),
+#'   order = exprs(AVISITN),
+#'   mode = "first",
+#'   keep_source_vars = exprs(AVISITN),
+#'   set_values_to = exprs(
+#'     PARAMCD = "ALK2",
+#'     PARAM = "ALKPH <= 2 times ULN"
+#'   )
+#' )
+#'
+#' # derive confirmed best overall response (using event_joined())
+#' adsl <- tribble(
+#'   ~USUBJID, ~TRTSDTC,
+#'   "1",      "2020-01-01",
+#'   "2",      "2019-12-12",
+#'   "3",      "2019-11-11",
+#'   "4",      "2019-12-30",
+#'   "5",      "2020-01-01",
+#'   "6",      "2020-02-02",
+#'   "7",      "2020-02-02",
+#'   "8",      "2020-02-01"
+#' ) %>%
+#'   mutate(TRTSDT = ymd(TRTSDTC))
+#'
+#' adrs <- tribble(
+#'   ~USUBJID, ~ADTC,        ~AVALC,
+#'   "1",      "2020-01-01", "PR",
+#'   "1",      "2020-02-01", "CR",
+#'   "1",      "2020-02-16", "NE",
+#'   "1",      "2020-03-01", "CR",
+#'   "1",      "2020-04-01", "SD",
+#'   "2",      "2020-01-01", "SD",
+#'   "2",      "2020-02-01", "PR",
+#'   "2",      "2020-03-01", "SD",
+#'   "2",      "2020-03-13", "CR",
+#'   "4",      "2020-01-01", "PR",
+#'   "4",      "2020-03-01", "NE",
+#'   "4",      "2020-04-01", "NE",
+#'   "4",      "2020-05-01", "PR",
+#'   "5",      "2020-01-01", "PR",
+#'   "5",      "2020-01-10", "PR",
+#'   "5",      "2020-01-20", "PR",
+#'   "6",      "2020-02-06", "PR",
+#'   "6",      "2020-02-16", "CR",
+#'   "6",      "2020-03-30", "PR",
+#'   "7",      "2020-02-06", "PR",
+#'   "7",      "2020-02-16", "CR",
+#'   "7",      "2020-04-01", "NE",
+#'   "8",      "2020-02-16", "PD"
+#' ) %>%
+#'   mutate(
+#'     ADT = ymd(ADTC),
+#'     PARAMCD = "OVR",
+#'     PARAM = "Overall Response by Investigator"
+#'   ) %>%
+#'   derive_vars_merged(
+#'     dataset_add = adsl,
+#'     by_vars = exprs(USUBJID),
+#'     new_vars = exprs(TRTSDT)
+#'   )
+#'
+#' derive_extreme_event(
+#'   adrs,
+#'   by_vars = exprs(USUBJID),
+#'   order = exprs(ADT),
+#'   mode = "first",
+#'   source_datasets = list(adsl = adsl),
+#'   events = list(
+#'     event_joined(
+#'       description = paste(
+#'         "CR needs to be confirmed by a second CR at least 28 days later",
+#'         "at most one NE is acceptable between the two assessments"
+#'       ),
+#'       join_vars = exprs(AVALC, ADT),
+#'       join_type = "after",
+#'       first_cond = AVALC.join == "CR" &
+#'         ADT.join >= ADT + 28,
+#'       condition = AVALC == "CR" &
+#'         all(AVALC.join %in% c("CR", "NE")) &
+#'         count_vals(var = AVALC.join, val = "NE") <= 1,
+#'       set_values_to = exprs(
+#'         AVALC = "CR"
+#'       )
+#'     ),
+#'     event_joined(
+#'       description = paste(
+#'         "CR needs to be confirmed by a second CR or PR at least 28 days later,",
+#'         "at most one NE is acceptable between the two assessments"
+#'       ),
+#'       join_vars = exprs(AVALC, ADT),
+#'       join_type = "after",
+#'       first_cond = AVALC.join %in% c("CR", "PR") &
+#'         ADT.join >= ADT + 28,
```

Comment:
![image](https://github.com/pharmaverse/admiral/assets/43018762/e420ee0d-bc4c-458a-afe5-56b437ca394b)

Should this not be a confirmed PR in the example?

Reply:
No, because between the two assessments all results must be CR, PR, or NE (and NE must occur only once).

Reply:
Gotcha, I thought SD's were allowed but I suppose its been a while since I've done it myself, always needed the chart/flow diagram to aid me anyhow